lib/cc/analyzer/include_paths_builder.rb

```diff
@@ -17,152 +14,76 @@ def initialize(cc_exclude_paths, cc_include_paths = [])
       end
 
       def build
-        if cc_include_paths.any?
-          paths = filter_by_cc_includes
-        else
-          root = Directory.new(".", ignored_files)
-          paths = root.included_paths
-        end
+        @_build ||= PathMinimizer.new(paths).minimize
+      end
 
-        paths.each do |path|
-          raise_on_unreadable_files(path)
-        end
+      def build_with_excludes(excludes)
+        new_paths = paths.reject { |path| matches_globs?(path, excludes) }
+        PathMinimizer.new(new_paths).minimize
       end
 
-      protected
+      private
 
-      def filter_by_cc_includes
-        paths = []
-        cc_include_paths.each do |path|
-          if File.directory?(path)
-            root = Directory.new(path, ignored_files)
-            paths += root.included_paths
-          elsif !ignored_files.include?(path)
-            paths << path
-          end
-        end
-        paths.uniq
+      def paths
+        @_paths ||=
+          all_matching_paths.
+          reject { |f| ignored_files.include?(f) }.
+          each { |f| raise_if_unreadable(f) }.
+          select { |f| FileUtils.readable_by_all?(f) }.
+          reject { |f| File.symlink?(f) }
       end
 
-      def ignored_files
-        Tempfile.open(".cc_gitignore") do |tmp|
-          tmp.write(File.read(".gitignore")) if File.file?(".gitignore")
-          tmp << @cc_exclude_paths.join("\n")
-          tmp.close
-          tracked_and_ignored = `git ls-files -zi -X #{tmp.path} 2>/dev/null`.split("\0")
-          untracked_and_ignored = `git ls-files -zio -X #{tmp.path} 2>/dev/null`.split("\0")
-          tracked_and_ignored + untracked_and_ignored
+      def all_matching_paths
+        include_paths.reject do |path|
+          matches_exclude?(path) || File.symlink?(path)
         end
       end
 
-      def raise_on_unreadable_files(path)
-        if File.directory?(path)
-          raise_on_unreadable_files_in_directory(path)
-        elsif !FileUtils.readable_by_all?(path)
-          raise CC::Analyzer::UnreadableFileError, "Can't read #{path}"
+      def include_paths
+        if @cc_include_paths.empty?
+          Dir.glob("**/*", File::FNM_DOTMATCH)
```

Comment:
I started looking into it but I don't think we'd want to call `glob_or_include_path` if `@cc_include_paths` is empty because `Dir.glob("**/*")` already includes all files / folders.

The case where we need `glob_or_include_path` is when the class is given `@cc_include_paths` that have directories since we need to work on the files in the directory at this stage.